### DISCUSSION TITLE
已经使用LoRa技术及进行了模型微调，产生了很多文件，如何使用微调后的模型呢？

### DISCUSSION THREAD

**Discussion by LiangYong1216** [OpenBMB/MiniCPM#97] [Q&A]
### Feature request  / 功能建议

使用LoRa技术微调，产生了下面的这些文件，有没有使用微调后的模型和原始模型一起进行预测的代码呢？
![image](https://github.com/OpenBMB/MiniCPM/assets/66230782/23a0b693-c0ae-4f0c-8923-e786029228af)

**Comment by Chen8566** [comment]
同问，使用SFT或者LORA的脚本训练后，生产出的模型需要怎么进行Chat或者推理？似乎没办法再用model.chat的方式了，是不是需要对输入做一些前置处理？

**Comment by LiangYong1216** [comment]
> 同问，使用SFT或者LORA的脚本训练后，生产出的模型需要怎么进行Chat或者推理？似乎没办法再用model.chat的方式了，是不是需要对输入做一些前置处理？

试试下面的代码：
import torch
from peft import PeftModel
from transformers import AutoModelForCausalLM, AutoTokenizer
path = '/mnt/ly/project/MiniCPM/models/MiniCPM'  # 替换你的基础模型路径
device = torch.device("cuda:2" if torch.cuda.is_available() else "cpu")  # 一张显卡就 cuda:0
# 加载基础模型
model = AutoModelForCausalLM.from_pretrained(path, torch_dtype=torch.float16, trust_remote_code=True).to(device)
# tokenizer编码
tokenizer = AutoTokenizer.from_pretrained(path)
# 加载PEFT模型（基础模型， 微调后的模型）
model = PeftModel.from_pretrained(model, '/mnt/ly/project/MiniCPM/finetune/output/AdvertiseGenLoRA/20240227091142/checkpoint-15000').to(device)

while True:
    content = input('请输入要查询的问题:\n')
    responds, history = model.chat(tokenizer, content, temperature=0.5, top_p=0.8, repetition_penalty=1.02, eos_token_id=2, pad_token_id=2)
    print(responds)
有结果了通知我一声，注意我打断点的地方，替换成你的模型路径：
![image](https://github.com/OpenBMB/MiniCPM/assets/66230782/39160797-661f-4131-b0bc-ee611bd4f15a)

**Comment by Chen8566** [comment]
> > 同问，使用SFT或者LORA的脚本训练后，生产出的模型需要怎么进行Chat或者推理？似乎没办法再用model.chat的方式了，是不是需要对输入做一些前置处理？
> 
> 试试下面的代码： import torch from peft import PeftModel from transformers import AutoModelForCausalLM, AutoTokenizer path = '/mnt/ly/project/MiniCPM/models/MiniCPM' # 替换你的基础模型路径 device = torch.device("cuda:2" if torch.cuda.is_available() else "cpu") # 一张显卡就 cuda:0
> 
> # 加载基础模型
> model = AutoModelForCausalLM.from_pretrained(path, torch_dtype=torch.float16, trust_remote_code=True).to(device)
> 
> # tokenizer编码
> tokenizer = AutoTokenizer.from_pretrained(path)
> 
> # 加载PEFT模型（基础模型， 微调后的模型）
> model = PeftModel.from_pretrained(model, '/mnt/ly/project/MiniCPM/finetune/output/AdvertiseGenLoRA/20240227091142/checkpoint-15000').to(device)
> 
> while True: content = input('请输入要查询的问题:\n') responds, history = model.chat(tokenizer, content, temperature=0.5, top_p=0.8, repetition_penalty=1.02, eos_token_id=2, pad_token_id=2) print(responds) 有结果了通知我一声，注意我打断点的地方，替换成你的模型路径： ![image](https://private-user-images.githubusercontent.com/66230782/308835470-39160797-661f-4131-b0bc-ee611bd4f15a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkxOTM5NjUsIm5iZiI6MTcwOTE5MzY2NSwicGF0aCI6Ii82NjIzMDc4Mi8zMDg4MzU0NzAtMzkxNjA3OTctNjYxZi00MTMxLWIwYmMtZWU2MTFiZDRmMTVhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI5VDA4MDEwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdmOTliY2IyMDZkMzU2OTM3NjFiMjA0YjI0YWFjOWMwOWZjMTU3ZjA2ZDIwN2NhYTRiZTU4YzgwZTdlZTBmYjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.R8zhqL4rYKXzfmaEuZNhE4z-HZ4hBDz5cEK11syNy80)

您好，按照您的代码可以加载模型，但是模型推理耗时非常非常久（可能到几分钟一条），我按照教程训练的广告生成，训练了15000 steps（8 epochs），但是效果极其不好。请问这是因为什么呀？
![image](https://github.com/OpenBMB/MiniCPM/assets/41323258/322f58fc-c0cf-42a7-aaba-b5ea0e4f8ce4)
![image](https://github.com/OpenBMB/MiniCPM/assets/41323258/780ee464-61de-4d0e-8a0f-f25b4436edd6)

**Comment by LiangYong1216** [comment]
15000step应该数据集一次还没迭代完吧？batch_size是1的话，并且使用一张卡。你的epoch=8是怎么算出来的？我现在也在微调，还没跑完，不知道结果怎么样呢。出来结果了我们可以对照一下

**Comment by LiangYong1216** [comment]
这是我训练的模型，输出也很烂，不知道是为什么？

![image](https://github.com/OpenBMB/MiniCPM/assets/66230782/45222976-557a-499a-8530-e2cacfbb9da5)

**Comment by Chen8566** [comment]
> 这是我训练的模型，输出也很烂，不知道是为什么？
> 
> ![image](https://private-user-images.githubusercontent.com/66230782/309649539-45222976-557a-499a-8530-e2cacfbb9da5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk1NzMyMzQsIm5iZiI6MTcwOTU3MjkzNCwicGF0aCI6Ii82NjIzMDc4Mi8zMDk2NDk1MzktNDUyMjI5NzYtNTU3YS00OTlhLTg1MzAtZTJjYWNmYmI5ZGE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA0VDE3MjIxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFkY2YxZWViNjViOGMwOTgzYjAzOGE0ODRhY2E3ZDY5OThhMGQzMzRjNGU0OGE1NmY0NThkMDg3MmM4ZDc2OWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.6ryTys4O4OVYYKsEPlxJIWFmv35EXbF1G0w0zRfi2JA)

我是4张卡，--per_device_train_batch_size 2，lora的训练效果很差。

**Comment by LiangYong1216** [comment]
有没有试过全量微调啊，用float32的话大概需要多少显存？

**Reply by jctime** [reply]
12G左右吧
